src/test/rpc/ShardArchiveHandler_test.cpp

```diff
@@ -0,0 +1,364 @@
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/beast/utility/temp_dir.h>
+#include <ripple/core/ConfigSections.h>
+#include <ripple/nodestore/DummyScheduler.h>
+#include <ripple/nodestore/Manager.h>
+#include <ripple/nodestore/impl/DecodedBlob.h>
+#include <ripple/protocol/jss.h>
+#include <ripple/rpc/ShardArchiveHandler.h>
+#include <test/jtx/Env.h>
+#include <test/jtx/TrustedPublisherServer.h>
+#include <test/jtx/envconfig.h>
+#include <test/nodestore/TestBase.h>
+
+namespace ripple {
+namespace test {
+
+class ShardArchiveHandler_test : public beast::unit_test::suite
+{
+    using Downloads = std::vector<std::pair<uint32_t, std::string>>;
+
+    TrustedPublisherServer
+    createServer(jtx::Env& env, bool ssl = true)
+    {
+        std::vector<TrustedPublisherServer::Validator> list;
+        list.push_back(TrustedPublisherServer::randomValidator());
+        return TrustedPublisherServer{
+            env.app().getIOService(),
+            list,
+            env.timeKeeper().now() + std::chrono::seconds{3600},
+            ssl};
+    }
+
+public:
+    void
+    testStateDatabase1()
+    {
+        testcase("testStateDatabase1");
+
+        {
+            beast::temp_dir tempDir;
+
+            auto c = jtx::envconfig();
+            auto& section = c->section(ConfigSection::shardDatabase());
+            section.set("path", tempDir.path());
+            section.set("max_size_gb", "100");
+            c->setupControl(true, true, true);
+
+            jtx::Env env(*this, std::move(c));
+            auto handler = RPC::ShardArchiveHandler::getInstance(
+                env.app(), env.app().getJobQueue());
+            BEAST_EXPECT(handler);
+
+            BEAST_EXPECT(handler->init());
+
+            std::string const rawUrl = "https://foo:443/1.tar.lz4";
+            parsedURL url;
+
+            parseUrl(url, rawUrl);
+            handler->add(1, {url, rawUrl});
+
+            {
+                std::lock_guard<std::mutex> lock(handler->m_);
+
+                auto& session{handler->sqliteDB_->getSession()};
+
+                soci::rowset<soci::row> rs =
+                    (session.prepare << "SELECT * FROM State;");
+
+                uint64_t rowCount = 0;
+
+                for (auto it = rs.begin(); it != rs.end(); ++it, ++rowCount)
+                {
+                    BEAST_EXPECT(it->get<int>(0) == 1);
+                    BEAST_EXPECT(it->get<std::string>(1) == rawUrl);
+                }
+
+                BEAST_EXPECT(rowCount == 1);
+            }
+
+            handler->release();
+        }
+
+        // Destroy the singleton so we start fresh in
+        // the next testcase.
+        RPC::ShardArchiveHandler::instance_.reset();
```

Comment:
It's unusual, but not unheard of, for a test to be a friend of the class it it testing.  That's the situation you have set up.  Since you have access to the `ShardArchiveHandler` internals, you could consider simply constructing a local `ShardArchiveHandler` in this test.  Then you wouldn't need to manually nuke the `instance_`.

Or, if you follow the suggestion of making the `ShardArchiveHandler` a member of the `Application`, then you can let the `Application` member of `Env` manage the lifetime for you.

This comment applies to all of the tests except, possibly, `testStateDatabase4()`.

Reply:
Went with the latter